raft_test.go

```diff
@@ -1019,6 +1020,88 @@ func TestRaft_SnapshotRestore(t *testing.T) {
 	}
 }
 
+func TestRaft_SnapshotRestore_Monotonic(t *testing.T) {
```

Comment:
Do we need this test anymore? I thought this was testing that a restore during NewRaft reset the log indexes, but we're no longer modifying that code path. I may be mistaken, but I think we might want to add [a check ](https://github.com/hashicorp/raft/pull/545#discussion_r1134154086) in the `snapShotAndRestore` helper, rather than here.

Reply:
IMO it's worth keeping if only to ensure we don't regress and accidentally re-add any trimming of logs to this code path in the future!

The fact that we could have added that code that would loose committed data but no existing tests would have caught it makes me feel like this is worth keeping for future!

I think we learned and made the tests better from this PR! Thanks Mike!

Reply:
I added https://github.com/hashicorp/raft/pull/545/commits/fc09cbd0a778414d990a44d3ed3aad45a0313a72 to verify whether logs are removed or not during user restores (and based off @mpalmi 's suggestion).  
I also added a check in here for when restoring to a new cluster vs the existing cluster.  I think it may also serve as the verification in raft of this integration test that is added to Consul: https://github.com/hashicorp/consul/pull/16647/files

Feel free to comment, adjust, or remove.

cc: @dhiaayachi @banks

Reply:
Great job @jmurret your new test that exercises all the combinations covers the actual snapshot Restore cases that we _did_ change.

I'd vote to keep this test for now since it validates that the "restore on startup" case does _not_ delete logs. Perhaps the one suggestion would be that we rename it to something like `TestRaft_RestoreSnapshotOnStartup_Monotonic` or something? TO make it clear that it's different from testing the `raft.RestoreSnapshot()` cases in the other test?

Reply:
Renamed this test.  Resolving.